timemachine/cpp/src/random_sampler.hpp

```diff
@@ -0,0 +1,38 @@
+#pragma once
+
+#include <memory>
+#include <vector>
+
+#include "curand.h"
+#include "device_buffer.hpp"
+#include <cub/util_type.cuh>
+
+namespace timemachine {
+
+// An RandomSampler for random sampling from probabilities (np.random.choice(X, replace=True p=probabilities))
+template <typename RealType> class RandomSampler {
+
+private:
+    const int N_; // Max number of values that can be sampled from
+    size_t temp_storage_bytes_;
+
+    DeviceBuffer<RealType> d_rand_;
```

Comment:
nit: `d_rand_` and `d_gumbel_` can probably share the same buffer since it looks like a `+=` operation.